docs/proposals/container-runtime-interface/capabilities.md

```diff
@@ -0,0 +1,134 @@
+to the [k8s API](https://github.com/kubernetes/kubernetes/blob/v1.5.0-alpha.0/pkg/api/v1/types.go#L1164-L1170).
+
+If a user wants to launch a container with required capabilities, he will go through such steps:
+
+- Get a list of capabilities he would require
+- Look up Docker's docs about the [default capability list](https://docs.docker.com/engine/reference/run/#/runtime-privilege-and-linux-capabilities)
+- Calculate the `Add` and `Drop` capability set based on the his requirement and the default capabilities list.
+- Specify the `Add` and `Drop` set for the containers in a k8s pod spec.
+- Launch the pod/container.
+
+There are several issues here:
+
+- The default capability set heavily depends on Docker's implementation, Kubernetes has no control over that.
+When Docker changes the default capability set across versions (hopefull it should not happen), users of Dockers get surprise,
+users of Kubernetes get surprise, too.
+- Even worse, to implement the same semantics as today's Docker runtime, it will require other runtimes (e.g. rkt) to depend
+on the default capability set as well. And if Docker changes the default capability set, then consistency is not guaranteed.
+- It is arguable that `Add` and `Drop` set are the best semantics for capability set compared to a [whitelist based](https://github.com/appc/spec/blob/af31bda9a474bf3a3e83144f6f0d264cfedc813f/spec/ace.md#oslinuxcapabilities-retain-set) semantics,
+which explicitly sets the bounding capabilities that a container could have, and has less security implications (users don't guess what
+is the final capability set).
+
+## Solutions
+
+As mentioned above, an explicit whitelist type of capability-retain-set could make things very consistent across runtimes and
+version updates.
+But as we can see, it's not very practical for system admins and operators to put a long [default capability list](https://docs.docker.com/engine/reference/run/#/runtime-privilege-and-linux-capabilities) on every
+container he wants to launch.
+Besides, changing the k8s API could take several release cycles.
+
+So in the proposed solution here. We will:
+- Keep the k8s API as is today. Users will not need to change anything.
+- Define a default capability list inside k8s, and document about it.
+- Make CRI expose the whitelist based capability interface, and put the burden on the runtime to implement that.
+- Make the final capability set **visible** through the container status, for example:
+
+```shell
+$ kubectl get pod -o yaml
+apiVersion: v1
+items:
+- apiVersion: v1
+  kind: Pod
+  metadata:
+  ...
+  spec:
+    containers:
+    - command:
+      - sleep
+      - "1000"
+      image: busybox
+      imagePullPolicy: Always
+      name: my-busybox
+      securityContext:
+        capabilities:
+          add:
+          - NET_ADMIN
+      ...
+  status:
+    containerStatuses:
+    - image: busybox
+    capabilities:
+      SETPCAP, MKNOD, AUDIT_WRITE, ...
```

Comment:
Almost. I think the logic looks like this (pseudo code):
```
if pod whitelist not set:
  pod whitelist = PSP default whitelist

add pod AddCapabilities to pod whitelist
remove pod DropCapabilities from pod whitelist

pass whitelist to runtime
```

- Ignore what I said before about validation
- `kubectl get pod -o yaml` shows the same whitelist as is passed to the runtime (resolved), plus add and drop

Since the PSP is still an optional component, I think we need to add in some additional defaulting in the Kubelet, but this is the basic idea.